packages/@atjson/editor/src/index.ts

```diff
@@ -1,184 +1,168 @@
-export default class Editor extends events(HTMLElement) {
-  static template = '<text-selection><div class="editor" contenteditable></div></text-selection><hr><div class="output"></div>';
-  static events = {
-    'beforeinput': 'beforeinput',
-    'change text-selection': 'updateSelection'
   };
 
-  text({ text }) {
-    return document.createTextNode(text);
+  get value() {
+    return this.document;
   }
 
-  paragraph() {
-    return document.createElement('p');
+  scheduleRender() {
+    window.requestAnimationFrame(() => {
+      this.render(this.querySelector('.editor'));
+      let evt = new CustomEvent('change', { bubbles: true, detail: { document: this.document } });
+      this.dispatchEvent(evt);
+    });
   }
 
-  bold() {
-    return document.createElement('strong');
-  }
+  render(editor) {
+    let rendered = new WebComponentRenderer(this.document).render();
 
-  italic() {
-    return document.createElement('em');
-  }
+    // This can be improved by doing the comparison on an element-by-element
+    // basis (or by rendering incrementally via the HIR), but for now this will
+    // prevent flickering of OS UI elements (e.g., spell check) while typing
+    // characters that don't result in changes outside of text elements.
+    if (rendered.innerHTML != editor.innerHTML) {
+      editor.innerHTML = rendered.innerHTML;
 
-  'line-break'() {
-    return document.createElement('br');
+      if (this.selection) {
+        this.querySelector('text-selection').setSelection(this.selection, { suppressEvents: true });
+      }
+    }
   }
 
-  beforeinput(evt) {
-    let ranges = evt.getTargetRanges();
-    let editor = this.querySelector('.editor');
-
-    let base = getNodeAndOffset(ranges[0].startContainer, ranges[0].startOffset);
-    let extent = getNodeAndOffset(ranges[0].endContainer, ranges[0].endOffset);
-    let start = editor.hir.get(base.node).start + base.offset;
-    let end = editor.hir.get(extent.node).end + extent.offset;
-
-    switch (evt.inputType) {
-    case 'insertText':
-      this.document.insertText(start, evt.data);
-      break;
-    case 'insertLineBreak':
-      this.document.insertText(start, '\u2028', true);
-      this.document.addAnnotations({
-        type: 'line-break',
-        start: start,
-        end: end + 1
-      });
-      break;
-    case 'formatBold':
-      this.document.addAnnotations({
-        type: 'bold',
-        start,
-        end
-      });
-    case 'insertParagraph':
-      this.document.insertText(start, '\n', true);
-      this.document.addAnnotations({
-        type: 'paragraph',
-        start: end + 1,
-        end: end + 1
-      });
-      break;
-    case 'deleteContentBackward':
-    case 'deleteContentForward':
-      this.document.deleteText({ start, end });
-      break;
-    }
+  setDocument(value: Document) {
+    this.document = value;
 
-    this.render(this.querySelector('.output'));
-  }
+    // n.b., would be good to have a way to query for existence of id on
```

Comment:
Probably will prefer making ids required in the future; that'll be considerably easier to deal with.

Reply:
👍